python/src/array.cpp

```diff
@@ -168,52 +186,92 @@ void get_shape(T list, std::vector<int>& shape) {
       return get_shape(l.template cast<py::list>(), shape);
     } else if (py::isinstance<py::tuple>(l)) {
       return get_shape(l.template cast<py::tuple>(), shape);
+    } else if (py::isinstance<array>(l)) {
+      auto arr = py::cast<array>(l);
+      shape.insert(shape.end(), arr.shape().begin(), arr.shape().end());
+      return;
     }
   }
 }
 
-template <typename T>
-array array_from_list(T pl, std::optional<Dtype> dtype) {
-  // Compute the shape
-  std::vector<int> shape;
-  get_shape(pl, shape);
-
-  // Validate the shape and type
-  auto type = validate_shape(pl, shape, 0);
+using array_init_type = std::variant<
+    py::bool_,
+    py::int_,
+    py::float_,
+    std::complex<float>,
+    py::list,
+    py::tuple,
+    py::array,
+    py::buffer,
+    py::object>;
 
-  size_t size = 1;
-  for (auto s : shape) {
-    size *= s;
-  }
+// Forward declaration
+array create_array(array_init_type v, std::optional<Dtype> t);
 
+template <typename T>
+array array_from_list(
+    T pl,
+    const Dtype& inferred_type,
+    std::optional<Dtype> specified_type,
+    const std::vector<int>& shape) {
   // Make the array
-  switch (type) {
-    case pybool: {
+  switch (inferred_type) {
+    case bool_: {
       std::vector<bool> vals;
       fill_vector(pl, vals);
-      return array(vals.begin(), shape, dtype.value_or(bool_));
+      return array(vals.begin(), shape, specified_type.value_or(inferred_type));
     }
-    case pyint: {
+    case int32: {
       std::vector<int> vals;
       fill_vector(pl, vals);
-      return array(vals.begin(), shape, dtype.value_or(int32));
+      return array(vals.begin(), shape, specified_type.value_or(inferred_type));
     }
-    case pyfloat: {
+    case float32: {
       std::vector<float> vals;
       fill_vector(pl, vals);
-      return array(vals.begin(), shape, dtype.value_or(float32));
+      return array(vals.begin(), shape, specified_type.value_or(inferred_type));
     }
-    case pycomplex: {
+    case complex64: {
       std::vector<std::complex<float>> vals;
       fill_vector(pl, vals);
       return array(
           reinterpret_cast<complex64_t*>(vals.data()),
           shape,
-          dtype.value_or(complex64));
+          specified_type.value_or(inferred_type));
+    }
+    default: {
+      std::ostringstream msg;
+      msg << "Should not happen, inferred: " << inferred_type
+          << " on subarray made of only python primitive types.";
+      throw std::runtime_error(msg.str());
     }
   }
 }
 
+template <typename T>
+array array_from_list(T pl, std::optional<Dtype> dtype) {
+  // Compute the shape
+  std::vector<int> shape;
+  get_shape(pl, shape);
+
+  // Validate the shape and type
+  bool all_python_primitive_elements = true;
+  auto type = validate_shape(pl, shape, 0, all_python_primitive_elements);
+
+  if (all_python_primitive_elements) {
+    // `pl` does not contain mlx arrays
+    return array_from_list(pl, type, dtype, shape);
```

Comment:
This code path handles only the code path when all elements are python primitive types